### ISSUE TITLE
32 bit version?

### ISSUE THREAD

**Issue by Halfshadow** [TASEmulators/desmume#872]
## State your operating system:
Windows 7 32 bit

## DesMuME version
0.9.13

## Isse
I want to download it for 32 bit Windows 7, but also if your website:
https://www.desmume.com/download.htm
Indicate it in the link like compatible with both 64 bit and 32 bit, the link indicated like for the 0.9.13 version it's a download instead for the 0.9.11 version... You could tell me that my OS is outdated and I could agree with you, but why you told that was available for 32 bit Windows 7 if isn't really?

Otherwise if is only a mistake, could you please then release the 32 bit version?

**Comment by zeromus** [comment]
desmume.com is a pirate website. We don't run it. We don't support 32bit windows anymore.

**Comment by rofl0r** [comment]
@Halfshadow i needed a 32 bit build too, so i made one using mingw-w64. unfortunately it crashes when selecting jit (at least with wine) and when recording an avi, but otherwise it seems to work well. i'd like your feedback whether jit works on "real windows" or whether it crashes there too.
[desmume.zip](https://github.com/user-attachments/files/18031287/desmume.zip)

**Comment by Halfshadow** [comment]
Hi. Thank you, but exactly like for you on Wine it crashes for me too during AVI recording and setting the Dynamic recompiler and then it plays slow. I need of an updated version 'cause the 0.9.11 starts ok, but after 5 minutes become unstable and very slow stressing the computer fan, instead I can execute No$GBA with the disaster frame skip without never have problems and without stress the CPU.

But I need of Desmume 'cause I need of the hokeys to set the frame skip only when I want without have to set it in boring way by the GUI but only pressing a button, and 'cause No$GBA doesn't allow to decrease the volume (is too high and I need to unmute the PC) and with HG/SS I need everytime I change something in the rom with the hacking, add again and again the cheat codes to bypass the antipiracy.

I could use a buggy 32 bit version without the function to record a movie, but I need of a stable fast framerate... Could you please do something about?

**Comment by zeromus** [comment]
You can change the volume of every application in the windows volume mixer. You can change the volume of your speakers. I can't believe this is really a reason for needing desmume and then asking for a 32bit version. 

Why are you having to continually re-renter cheat codes? Does nocash not remember them?

**Comment by Halfshadow** [comment]
Nocash remember the cheat codes but don't reload automatically and I didn't found an option to set it and recognize the rom when I add some scripts (I'm modding Soul Silver) and then the rom change the hash and the same cheat code need to be readded again 'cause not only cannot reload the cheat code, but obviously cannot reload a common cheat code for every rom, there isn't an option. In Desmume the cheat code isn't necessary.

Don't support not even savestates, are useful to check the modded script in the game with the variables.

Nocash also don't have the hotkeys function to remove the framerate limit only pressing a button to release it when necessary, then I don't need of a Desmume able to record videos, I can use the 0.9.11 to record videos considering that would become slow however, but I need of a Desmume 0.9.11 that cannot stress the CPU or of a 0.9.13 where I could use the framerate limiter bypass with a joypad button (the "fast forward" hotkey association in Desmume not existing in nocash) the modded 32 bit exe that rofl0r passed me works for these things I need, except for the Dynamic ricompiler, that could help to fix this issue, otherwise I could use 0.9.11 if could works without stress the CPU, it does it also if I don't open no roms but simply is opened... Checking it from task manager in await to load a rom costs 20% of the CPU... The 0.9.13 during the await doesn't do this and indeed I thonk could works very fine with the dynamic recompiler...

**Comment by zeromus** [comment]
ok that was a lot of text. if you're hacking but need desmume then I guess that's cool. as for "stress cpu", sorry, desmume is for gamers. set cpu priority to low via windows tools or change display method and 3d renderer to softrasterizer. asking desmume to be "low cpu" is simply not going to happen now. don't bother dreaming. maybe try the x432r version

**Comment by rofl0r** [comment]
@Halfshadow so you're saying the dynamic jit recompiler in my build works, but slower than 0.9.11 ?

**Comment by Halfshadow** [comment]
@zeromus Was already to Softrasterizer and set the priority to low didn't work.
I see what you mean but reading on the web the 0.9.13 solved the bug of the CPU stress and considering that Nocash had no these problems I could think to solve with a version rivisited.

Indeed X432R works better by 0.9.11 and own a CPU power saving option and I can ability the dynamic ricompiler. I simply didn't know was existing, I asked before if there were alternatives but nobody told me about. Thank you for this then, certainly works better.

**Comment by Halfshadow** [comment]
@rofl0r nope, doesn't work unfortunately. Thank you for your help, but I think isn't possible...
However if X432R works I think it's enough, I need of speed hotkeys, savestates and load HG/SS roms without insert the cheat codes to bypass the antipiracy everytime, if this works I think is enough... Otherwise if you want to try again I'll test it for you. Sure.

**Comment by rofl0r** [comment]
i managed to fix both bugs, this version should be as fast as 32bit desmume can get: 
[desmume.zip](https://github.com/user-attachments/files/18042906/desmume.zip)

**Comment by Halfshadow** [comment]
CA 300 in Pokémon Soul Silver during a battle and CA 500 in the same game and same condition but with X432R version...

Without lags with the Softrasterizer plugin, instead in the 0.9.14 32 bit mod during instead the walking in the maps I see lags without during the normal execution, with the FPS limiter.

Using the Softrasterizer in 32 bit mod 0.9.14 causes frame falls during simple walks.

**Comment by zeromus** [comment]
I can't decipher this. Perhaps summarize it in a table.

**Comment by Halfshadow** [comment]
@zeromus Sorry I forgot the tag. Was in reply to @rofl0r that asked me how many frames goes his 0.9.14 32 bit mod and the X432R fork.

**Comment by zeromus** [comment]
I understood that. But I cannot understand which framerates correspond to which tests in which builds of desmume in which games. I doubt rofl0r can either. Hence I suggested a table.

**Comment by rofl0r** [comment]
i figured that CA X means "approx X fps". meanwhile i compared perf of my build to 0.9.11 exe, and it's achieving only 50-70% of the performance, with jit making almost no difference. trying to figure out what's different in the 2 builds, after adding some march, mtune, mmx and sse2 flags i get ~20% better performance, but it's still far off. so i suspect the 0.9.11 build used some hidden defines to speed up 32 bit/win32 performance that aren't currently enabled, since gcc is known to generally produce better code than msvc. i also noticed that the jit crashes if i use any calling convention other than default and regparm3. even regparm2 crashes.

**Comment by Jules-A** [comment]
> so i suspect the 0.9.11 build used some hidden defines to speed up 32 bit/win32 performance that aren't currently enabled, since gcc is known to generally produce better code than msvc. 

Wasn't the 0.9.11 build built with PGO? I could have sworn when I downloaded it many years ago it included the profile files. I still have an old PGO build I did from 2019 that uses up to AVX and it's still ~20% faster than current day builds even though there's been many optimizations since then. If I can recall it was something like a 45% speedup at the time on my FX8320.

**Comment by rofl0r** [comment]
that's an interesting anecdote, and i tried to make a pgo build to collect profiling samples. but unfortunately, it crashes on startup:

```
Unhandled exception: page fault on read access to 0xf9377de4 in 32-bit code (0x010161c7).
Register dump:
 CS:0023 SS:002b DS:002b ES:002b FS:0063 GS:006b
 EIP:010161c7 ESP:0ce1fe4c EBP:0ce1fe58 EFLAGS:00010206(  R- --  I   - -P- )
 EAX:f7cc3700 EBX:00f0d749 ECX:00000000 EDX:3630352a
 ESI:0d810e50 EDI:00000007
Stack dump:
0x0ce1fe4c:  0000010d 0d810e50 00000007 0ce1fe78
0x0ce1fe5c:  00f0d76b 1c3de9b7 00000000 00f0d749
0x0ce1fe6c:  6a28f1b2 00000000 00000007 0ce1ff28
0x0ce1fe7c:  01017eba 0d810ed0 0013177b 00000007
0x0ce1fe8c:  0ce20000 00000003 00131738 00000004
0x0ce1fe9c:  00401357 0d810ed0 0013177b 00000007
Backtrace:
=>0 0x010161c7 __gcov_indirect_call_profiler_v2+0xf(value=0x1c3de9b7, cur_func=00F0D749) in desmume (0x0ce1fe58)
  1 0x00f0d76b _GLOBAL__sub_I_HudEditorMode+0x22() in desmume (0x0ce1fe78)
  2 0x01017eba __do_global_ctors+0x1a() in desmume (0x0ce1ff28)
  3 0x00401357 __tmainCRTStartup+0x207() in desmume (0x0ce1ff28)
  4 0x7b62de50 WriteTapemark+0x100(device=<is not available>, type=<is not available>, count=<is not available>, immediate=<is not available>) [Z:\root\wine-6.22\dlls\kernel32\tape.c:317] in kernel32 (0x0ce1ff48)
  5 0x7bc57a77 RtlWakeConditionVariable+0x57(variable=<is not available>) [Z:\root\wine-6.22\dlls\ntdll\sync.c:766] in ntdll (0x0ce1ff5c)
  6 0x7bc58130 RtlCreateUserThread(entry=00401490, arg=7FFD1000) [Z:\root\wine-6.22\dlls\ntdll\thread.c:261] in ntdll (0x0ce1ffec)
0x010161c7 __gcov_indirect_call_profiler_v2+0xf in desmume: cmpl	%ebx,0x16b46e4(%eax)
```
... and debugging gcc code on an alien platform isn't really my cup of tea.
while researching the subject, i've seen claims of up to 14% performance improvements, but that still wouldn't account in the difference i can see, unless desmume is a quite exceptional case.
in the zip i downloaded, there are no profile files included.
my primary interest in getting a win32 build working again is to compare features and menu item placement so when backporting those to gtk i can study the "reference implementation". that it's not quite on par performance-wise is annoying, but not that important for my usecase.
that said, if i get a surge of motivation in the coming week i might try to push this further, but if not i'm quite OK with the status of the win32 port as-is. once it can be built by others someone else might show up who can identify the issue behind the performance regression.

**Comment by Jules-A** [comment]
Yeah, I tried to create a PGO profile in MSVC to test what it would be like on my new CPU with newer optimizations but it would just crash after 7secs after launching the ROM so couldn't build a profile to test. I'm clueless when it comes to finding what's causing the crash though :/
